<a href="https://colab.research.google.com/github/Azmal16/Machine_Learning_Practice/blob/master/DL%20Books/Deep%20Learning%20with%20Python%20by%20Chollet/Chapter_2/Digit_Recognition_with_Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Importing Dataset
The MNIST dataset comes preloaded in Keras, in the form of a set of four Numpy arrays.

In [1]:
from keras.datasets import mnist

##Loading Dataset


In [2]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


##Looking at the Training Data:

In [3]:
train_images.shape

(60000, 28, 28)

In [4]:
len(train_labels)

60000

In [5]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

##Looking at the Test Data:

In [6]:
test_images.shape

(10000, 28, 28)

In [7]:
len(test_labels)

10000

In [8]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

##Building the Network
We will build the **Network** first and then we will feed it the Training Data (*train_images* and *train_labels*. The network will then learn to associate images and labels. Then we will let the Network predict from the *test_images* and then we will compare the result with the **Test data.**

In [9]:
from keras import models
from keras import layers

In [10]:
network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,))) 
network.add(layers.Dense(10, activation='softmax'))

##The Compilation Step
As part of the Compilation Step, we will get **three things** ready:
1. A loss function: How the network will be able to measure its performance on the training data, and thus how it will be able to steer itself in the right direction.
2. An optimizer: The mechanism through which the network will update itself based on the data it sees and its loss function.
3. Metrics to monitor during training and testing: Here, we’ll only care about accuracy (the fraction of the images that were correctly classified).

In [11]:
network.compile(optimizer='rmsprop', 
                loss='categorical_crossentropy',
                metrics=['accuracy'])


##Preparing the image data
We’ll preprocess the data by reshaping it into the shape the network expects and scaling it so that all values are in the [0, 1] interval.

In [12]:

train_images = train_images.reshape((60000, 28 * 28)) 
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28)) 
test_images = test_images.astype('float32') / 255


##Preparing the labels
We will now categorically encode the labels.

In [13]:
from keras.utils import to_categorical

In [14]:
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

##Fitting the model to its training data
We’ll now train the network, which in Keras is done via a call to the network’s fit method—we fit the model to its training data.

In [19]:
network.fit(train_images, train_labels, epochs=7, batch_size=128)

Epoch 1/7
469/469 [==============================] - 1s 3ms/step - loss: 0.0060 - accuracy: 0.9984
Epoch 2/7
469/469 [==============================] - 1s 3ms/step - loss: 0.0049 - accuracy: 0.9986
Epoch 3/7
469/469 [==============================] - 1s 3ms/step - loss: 0.0038 - accuracy: 0.9990
Epoch 4/7
469/469 [==============================] - 1s 3ms/step - loss: 0.0030 - accuracy: 0.9992
Epoch 5/7
469/469 [==============================] - 1s 3ms/step - loss: 0.0024 - accuracy: 0.9995
Epoch 6/7
469/469 [==============================] - 1s 3ms/step - loss: 0.0019 - accuracy: 0.9995
Epoch 7/7
469/469 [==============================] - 1s 3ms/step - loss: 0.0017 - accuracy: 0.9995


##Looking at the Accuracy 
Now, we will look at the Accuracy which our network has delivered.

In [20]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.0967 - accuracy: 0.9822
test_acc: 0.982200026512146


As we can see, our network has the accuracy of 98%. Which is pretty good.